nvflare/app_opt/pt/quantization/ada_quant.py

```diff
@@ -0,0 +1,115 @@
+    def __init__(self, weight: float = 0.01, compression: bool = True) -> None:
+        """Implements the ADAQUANT quantization scheme,
+            for further details refer to the paper https://arxiv.org/abs/2208.05174
+
+        Args:
+            weight: a hyperparameter for the trade-off between quantization size and error
+            compression: whether to compress the resulting integer quantized tensor
+
+        """
+
+        self.weight = weight
+        self.compression = compression
+
+    def quantize(self, values_tensor: torch.Tensor) -> tuple[Union[torch.Tensor, np.ndarray], dict]:
+        values_tensor = values_tensor.cpu()
+        old_values_tensor = values_tensor
+
+        old_tensor_shape = list(old_values_tensor.shape)
+        values_tensor = values_tensor.to(dtype=torch.float64).view(-1)
+        offset = self.get_offset(values_tensor)
+        values_tensor = values_tensor + offset
+        norm = values_tensor.abs().max()
+
+        quant_state = {"offset": offset}
+        res = self.get_number_of_quantization_levels(
+            element_size=old_values_tensor.element_size(), values_tensor=values_tensor
+        )
+        if res is None:
+            return old_values_tensor, {}
+        norm, quantization_level, new_dtype = res
+        if norm == 0.0:
+            return torch.tensor([0], dtype=torch.bool), quant_state | {
+                "tensor_shape": old_tensor_shape,
+            }
+
+        normalized_abs_tensor = values_tensor.abs() / norm
+        quantized_tensor = (normalized_abs_tensor * quantization_level).round().clamp(0, quantization_level)
+        quantized_tensor = quantized_tensor.numpy().astype(dtype=new_dtype)
+        if self.compression:
+            raw_bytes = quantized_tensor.tobytes()
+            compressed_bytes = bz2.compress(raw_bytes)
+            if len(compressed_bytes) < len(raw_bytes):
+                compressed_tensor = np.frombuffer(bz2.compress(quantized_tensor.tobytes()), dtype=np.uint8)
+                return torch.tensor([0], dtype=torch.bool), quant_state | {
+                    "compressed_tensor": compressed_tensor,
+                    "quantization_level": quantization_level,
+                    "new_dtype": str(new_dtype),
+                    "tensor_shape": old_tensor_shape,
+                    "norm": norm,
+                }
+        return quantized_tensor, quant_state | {
+            "quantization_level": quantization_level,
+            "tensor_shape": old_tensor_shape,
+            "norm": norm,
+        }
+
+    def dequantized(self, quantized_tensor: torch.Tensor, quant_state: dict) -> torch.Tensor:
+        offset = quant_state["offset"]
+        if "norm" not in quant_state:
+            return torch.zeros(quant_state["tensor_shape"], dtype=torch.float64) - offset
+        norm = quant_state["norm"]
+        if "compressed_tensor" in quant_state:
+            decompressed_tensor = bz2.decompress(quant_state["compressed_tensor"].tobytes())
```

Comment:
maybe we should remove the compress part, it can be an independent function for streaming that can be turned on for universal usage @holgerroth @yanchengnv what do you think?